core/drivers/pm/sam/pm_suspend.S

```diff
@@ -0,0 +1,516 @@
+/* SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause) */
+/*
+ * Copyright (c) 2021, Microchip
+ */
+
+#include <arm.h>
+#include <arm32_macros.S>
+#include <asm.S>
+#include "at91_pm.h"
+#include <at91_pmc.h>
+#include <drivers/sam/at91_ddr.h>
+#include <generated/pm-defines.h>
+
+#define	SRAMC_SELF_FRESH_ACTIVE		0x01
+#define	SRAMC_SELF_FRESH_EXIT		0x00
+
+pmc	.req	r0
+tmp1	.req	r4
+tmp2	.req	r5
+tmp3	.req	r6
+
+/*
+ * Wait until master clock is ready (after switching master clock source)
+ */
+	.macro wait_mckrdy
+1:	ldr	tmp1, [pmc, #AT91_PMC_SR]
+	tst	tmp1, #AT91_PMC_MCKRDY
+	beq	1b
+	.endm
+
+/*
+ * Wait until master oscillator has stabilized.
+ */
+	.macro wait_moscrdy
+1:	ldr	tmp1, [pmc, #AT91_PMC_SR]
+	tst	tmp1, #AT91_PMC_MOSCS
+	beq	1b
+	.endm
+
+/*
+ * Wait for main oscillator selection is done
+ */
+	.macro wait_moscsels
+1:	ldr	tmp1, [pmc, #AT91_PMC_SR]
+	tst	tmp1, #AT91_PMC_MOSCSELS
+	beq	1b
+	.endm
+
+/*
+ * Put the processor to enter the idle state
+ */
+	.macro at91_cpu_idle
+
+	mov	tmp1, #AT91_PMC_PCK
+	str	tmp1, [pmc, #AT91_PMC_SCDR]
+
+	dsb
+
+	wfi		@ Wait For Interrupt
+
+	.endm
+
+	.section .text.psci.suspend
+
+	.arm
+
+
+#define SUSPEND_FUNC(__name) \
+__name:
+
+#define SUSPEND_END_FUNC(__name) \
+	.size __name, .-__name
+
+.macro check_fit_in_sram since
+	.if (. - \since) > 0x10000
+		.error "Suspend assembly code exceeds dedicated SRAM size"
+	.endif
+.endm
+
+/*
+ * void at91_suspend_sram_fn(struct at91_pm_data*)
+ * @input param:
+ * 	@r0: base address of struct at91_pm_data
+ */
+	.align 3
+	.global at91_pm_suspend_in_sram
+SUSPEND_FUNC(at91_pm_suspend_in_sram)
```

Comment:
When using the` FUNC/END_FUNC` assembly macro, there is an error during compilation for the check at the end:
```
core/drivers/pm/sam/pm_suspend.S: Assembler messages:
core/drivers/pm/sam/pm_suspend.S:516: Error: non-constant expression in ".if" statement
core/drivers/pm/sam/pm_suspend.S:514: Error: can't resolve .text.at91_backup_mode - at91_pm_suspend_in_sram
```
It looks like the compiler can figure out the symbols name placement. I tried replacing at first with `FUNC` and `END_FUNC` but did not found a way to do it. Let me know if you have some trick to do it.

Reply:
Please check other examples of how these macros are used.

Reply:
Other examples (in `entry_a64.S`) uses the check macro inside the declaration of the symbol and inside the function check. If I do the same thing in my code, it works but its obviously not what I want.
I also tried adding a symbol before the FUNC declaration but it does not work either.
Using  `FUNC/END_FUNC` between the first symbol I uses (`at91_pm_suspend_in_sram`) and the check at the end yields the error I previously mentionned. I suspect this is due to the compiler not being able to resolve the symbols at compile time because your macro uses assembly macro and not preprocessor defines.
My question remains, I genuinely did not found a way to make it works, do you have any idea on how do do that using `FUNC/END_FUNC` ? Thanks.

Reply:
It seems to be more trouble than it's worth to use these macros here. Let's keep it as it is.